azurerm/internal/services/msi/migration/user_assigned_identity_V0_to_V1.go

```diff
@@ -0,0 +1,62 @@
+package migration
+
+import (
+	"log"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/tags"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/clients"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/msi/parse"
+)
+
+func UserAssignedIdentityV0ToV1() schema.StateUpgrader {
+	return schema.StateUpgrader{
+		Version: 0,
+		Type:    userAssignedIdentityV0Schema().CoreConfigSchema().ImpliedType(),
+		Upgrade: userAssignedIdentityUpgradeV0ToV1,
+	}
+}
+
+func userAssignedIdentityV0Schema() *schema.Resource {
+	return &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringLenBetween(3, 128),
+			},
+
+			"resource_group_name": azure.SchemaResourceGroupName(),
+
+			"location": azure.SchemaLocation(),
+
+			"tags": tags.Schema(),
+
+			"principal_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"client_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func userAssignedIdentityUpgradeV0ToV1(rawState map[string]interface{}, meta interface{}) (map[string]interface{}, error) {
+	subscriptionId := meta.(*clients.Client).Account.SubscriptionId
```

Comment:
there's one issue here where we assume the subscription being used is the same one used to provision the resource, which is the case 99% of the time, but may not be during state migrations if someone's using the Azure CLI and has updated their default) - as such instead it'd be worth parsing the existing ID which'll pull out the SubscriptionID being used:

```suggestion
	oldId := rawState["id"].(string)
	id, err := parse.UserAssignedIdentityID(oldId)
	if err != nil {
		return rawState, err
	}
	
	newId := id.ID()
	log.Printf("Updating `id` from %q to %q", oldId, newId)
	rawState["id"] = newId
	return rawState, nil
```

arguably there's a separate issue here insofar as we're not updating the Subscription used for the clients on demand to account for this, but that's a much larger issue we're not likely to fix anytime soon